IncompatibleClassChangeError from okhttp3.internal.Util.closeQuietly

Using 3.6.0-SNAPSHOT, we are seeing `IncompatibleClassChangeError` exceptions thrown from `okhttp3.internal.Util.closeQuietly()` on Android versions prior to 4.4. It can be reliably reproduced, but it's challenging to write a test, as we're only seeing the error consistently when the server socket closes unexpectedly. So I've described steps to reproduce here.

Here's the stack:
```
Caused by: java.lang.IncompatibleClassChangeError: interface not implemented
at okhttp3.internal.Util.closeQuietly(Util.java:100)
at okhttp3.internal.connection.StreamAllocation.streamFailed(StreamAllocation.java:332)
at okhttp3.internal.http.RetryAndFollowUpInterceptor.recover(RetryAndFollowUpInterceptor.java:209)
at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:132)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:212)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:179)
at okhttp3.RealCall.execute(RealCall.java:63)
```

To reproduce the problem, we hosted the following php (borrowed from [http://stackoverflow.com/a/141026/315702](http://stackoverflow.com/a/141026/315702)) on a server:

```
<?php
ob_end_clean();
header("Connection: close");
ignore_user_abort(true); // just to be safe
ob_start();
echo('Text the user will see');
$size = ob_get_length();
header("Content-Length: $size");
ob_end_flush(); // Strange behaviour, will not work
flush(); // Unless both are called !
sleep(30);
echo('Text user will never see');
?>
```

This is a boiled-down version of our okhttp call:

```
Context context = aParameter;
String body = anotherParameter;
File cacheDirectory = context.getDir("ts_okhttp_cache", Context.MODE_PRIVATE);
int cacheSize = 5 * 1024 * 1024; // 5 MiB
Cache fileCache = new Cache(cacheDirectory, cacheSize);
OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder();
clientBuilder.cache(fileCache);
OkHttpClient client = clientBuilder.build();
MediaType jsonMediaType = MediaType.parse("application/json; charset=utf-8");
Request.Builder requestBuilder = new Request.Builder();
requestBuilder.url(urlString);
requestBuilder.post(RequestBody.create(jsonMediaType, body));
return client.newCall(requestBuilder.build()).execute();
```

Every time we make a call to that php page, using Android emulators running Android 4.3 or earlier, we get the `IncompatibleClassChangeError` exception. Just guessing, but it seems like this could be due to a lack of bytecode backward compatibility with Java 6, where `java.net.Socket` doesn't implement `Closeable`.

When we switch to okhttp 3.5.0, the problem goes away. (Same goes for 3.4.1.)

A note for any other developers who have run into this and are puzzled, because they don't think they're using 3.6.0-SNAPSHOT... make sure to check the dependencies of your app. You probably have a third-party library that's pulling in okhttp:+. In our case, it was the okhttp extension for exoplayer, and we were able to block the dependency with the following change to our build.gradle file:

```
compile('com.google.android.exoplayer:extension-okhttp:r2.0.4') {
  exclude module: 'okhttp'
}
```

--- a/okhttp/src/main/java/okhttp3/ConnectionPool.java
+++ b/okhttp/src/main/java/okhttp3/ConnectionPool.java
@@ -16,8 +16,8 @@
  */
 package okhttp3;
 
-import java.io.Closeable;
 import java.lang.ref.Reference;
+import java.net.Socket;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Deque;
@@ -130,7 +130,7 @@ public final class ConnectionPool {
    * Replaces the connection held by {@code streamAllocation} with a shared connection if possible.
    * This recovers when multiple multiplexed connections are created concurrently.
    */
-  Closeable deduplicate(Address address, StreamAllocation streamAllocation) {
+  Socket deduplicate(Address address, StreamAllocation streamAllocation) {
     assert (Thread.holdsLock(this));
     for (RealConnection connection : connections) {
       if (connection.isEligible(address)

--- a/okhttp/src/main/java/okhttp3/OkHttpClient.java
+++ b/okhttp/src/main/java/okhttp3/OkHttpClient.java
@@ -15,10 +15,10 @@
  */
 package okhttp3;
 
-import java.io.Closeable;
 import java.net.MalformedURLException;
 import java.net.Proxy;
 import java.net.ProxySelector;
+import java.net.Socket;
 import java.net.UnknownHostException;
 import java.security.GeneralSecurityException;
 import java.security.KeyStore;
@@ -149,7 +149,7 @@ public class OkHttpClient implements Cloneable, Call.Factory, WebSocket.Factory 
         return pool.get(address, streamAllocation);
       }
 
-      @Override public Closeable deduplicate(
+      @Override public Socket deduplicate(
           ConnectionPool pool, Address address, StreamAllocation streamAllocation) {
         return pool.deduplicate(address, streamAllocation);
       }

--- a/okhttp/src/main/java/okhttp3/internal/Internal.java
+++ b/okhttp/src/main/java/okhttp3/internal/Internal.java
@@ -15,8 +15,8 @@
  */
 package okhttp3.internal;
 
-import java.io.Closeable;
 import java.net.MalformedURLException;
+import java.net.Socket;
 import java.net.UnknownHostException;
 import javax.net.ssl.SSLSocket;
 import okhttp3.Address;
@@ -55,7 +55,7 @@ public abstract class Internal {
   public abstract RealConnection get(
       ConnectionPool pool, Address address, StreamAllocation streamAllocation);
 
-  public abstract Closeable deduplicate(
+  public abstract Socket deduplicate(
       ConnectionPool pool, Address address, StreamAllocation streamAllocation);
 
   public abstract void put(ConnectionPool pool, RealConnection connection);

--- a/okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java
+++ b/okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java
@@ -15,10 +15,10 @@
  */
 package okhttp3.internal.connection;
 
-import java.io.Closeable;
 import java.io.IOException;
 import java.lang.ref.Reference;
 import java.lang.ref.WeakReference;
+import java.net.Socket;
 import okhttp3.Address;
 import okhttp3.ConnectionPool;
 import okhttp3.OkHttpClient;
@@ -186,7 +186,7 @@ public final class StreamAllocation {
     result.connect(connectTimeout, readTimeout, writeTimeout, connectionRetryEnabled);
     routeDatabase().connected(result.route());
 
-    Closeable closeable = null;
+    Socket socket = null;
     synchronized (connectionPool) {
       // Pool the connection.
       Internal.instance.put(connectionPool, result);
@@ -194,17 +194,17 @@ public final class StreamAllocation {
       // If another multiplexed connection to the same address was created concurrently, then
       // release this connection and acquire that one.
       if (result.isMultiplexed()) {
-        closeable = Internal.instance.deduplicate(connectionPool, address, this);
+        socket = Internal.instance.deduplicate(connectionPool, address, this);
         result = connection;
       }
     }
-    closeQuietly(closeable);
+    closeQuietly(socket);
 
     return result;
   }
 
   public void streamFinished(boolean noNewStreams, HttpCodec codec) {
-    Closeable closeable;
+    Socket socket;
     synchronized (connectionPool) {
       if (codec == null || codec != this.codec) {
         throw new IllegalStateException("expected " + this.codec + " but was " + codec);
@@ -212,9 +212,9 @@ public final class StreamAllocation {
       if (!noNewStreams) {
         connection.successCount++;
       }
-      closeable = deallocate(noNewStreams, false, true);
+      socket = deallocate(noNewStreams, false, true);
     }
-    closeQuietly(closeable);
+    closeQuietly(socket);
   }
 
   public HttpCodec codec() {
@@ -232,20 +232,20 @@ public final class StreamAllocation {
   }
 
   public void release() {
-    Closeable closeable;
+    Socket socket;
     synchronized (connectionPool) {
-      closeable = deallocate(false, true, false);
+      socket = deallocate(false, true, false);
     }
-    closeQuietly(closeable);
+    closeQuietly(socket);
   }
 
   /** Forbid new streams from being created on the connection that hosts this allocation. */
   public void noNewStreams() {
-    Closeable closeable;
+    Socket socket;
     synchronized (connectionPool) {
-      closeable = deallocate(true, false, false);
+      socket = deallocate(true, false, false);
     }
-    closeQuietly(closeable);
+    closeQuietly(socket);
   }
 
   /**
@@ -255,7 +255,7 @@ public final class StreamAllocation {
    * <p>Returns a closeable that the caller should pass to {@link Util#closeQuietly} upon completion
    * of the synchronized block. (We don't do I/O while synchronized on the connection pool.)
    */
-  private Closeable deallocate(boolean noNewStreams, boolean released, boolean streamFinished) {
+  private Socket deallocate(boolean noNewStreams, boolean released, boolean streamFinished) {
     assert (Thread.holdsLock(connectionPool));
 
     if (streamFinished) {
@@ -264,7 +264,7 @@ public final class StreamAllocation {
     if (released) {
       this.released = true;
     }
-    Closeable closeable = null;
+    Socket socket = null;
     if (connection != null) {
       if (noNewStreams) {
         connection.noNewStreams = true;
@@ -274,13 +274,13 @@ public final class StreamAllocation {
         if (connection.allocations.isEmpty()) {
           connection.idleAtNanos = System.nanoTime();
           if (Internal.instance.connectionBecameIdle(connectionPool, connection)) {
-            closeable = connection.socket();
+            socket = connection.socket();
           }
         }
         connection = null;
       }
     }
-    return closeable;
+    return socket;
   }
 
   public void cancel() {
@@ -299,7 +299,7 @@ public final class StreamAllocation {
   }
 
   public void streamFailed(IOException e) {
-    Closeable closeable;
+    Socket socket;
     boolean noNewStreams = false;
 
     synchronized (connectionPool) {
@@ -326,10 +326,10 @@ public final class StreamAllocation {
           route = null;
         }
       }
-      closeable = deallocate(noNewStreams, false, true);
+      socket = deallocate(noNewStreams, false, true);
     }
 
-    closeQuietly(closeable);
+    closeQuietly(socket);
   }
 
   /**
@@ -364,19 +364,19 @@ public final class StreamAllocation {
    * <p>Returns a closeable that the caller should pass to {@link Util#closeQuietly} upon completion
    * of the synchronized block. (We don't do I/O while synchronized on the connection pool.)
    */
-  public Closeable releaseAndAcquire(RealConnection newConnection) {
+  public Socket releaseAndAcquire(RealConnection newConnection) {
     assert (Thread.holdsLock(connectionPool));
     if (codec != null || connection.allocations.size() != 1) throw new IllegalStateException();
 
     // Release the old connection.
     Reference<StreamAllocation> onlyAllocation = connection.allocations.get(0);
-    Closeable closeable = deallocate(true, false, false);
+    Socket socket = deallocate(true, false, false);
 
     // Acquire the new connection.
     this.connection = newConnection;
     newConnection.allocations.add(onlyAllocation);
 
-    return closeable;
+    return socket;
   }
 
   public boolean hasMoreRoutes() {

